introcude SCMI agent driver

etienne-lms opened the pull request:
Introduce a SMCI agent driver and support for SCMI clock and SCMI reset domain protocol and device drivers.
With this change, SCMI agent can support 2 possible transports layers: a mailbox device based and an Arm SMC invocation based.A former change can enhance the driver to support a SMCCC conduit driver to support Arm HVC invocation path.

apalos commented on drivers/clk/clk_scmi.c:
```diff
@@ -0,0 +1,149 @@
+{
+	struct scmi_clk_rate_set_in in = {
+		.clock_id = clk->id,
+		.flags = SCMI_CLK_RATE_ASYNC_NORESP | \
```
No need for \

etienne-lms replied:
ok

apalos commented on include/scmi_agent.h:
```diff
@@ -0,0 +1,86 @@
+{
+	if (errno <= SCMI_SUCCESS && errno > SCMI_ERR_MAX)
+		return scmi_linux_errmap[-errno];
+	return -EIO;
```
Is this only supposed to be used in I/O related functions? If not maybe EINVAL is a better default return option?

etienne-lms replied:
I tought `EIO` would highlight that there was some communication error, as SCMI is related to communication between 2 firmwares.

apalos commented on include/scmi_agent.h:
```diff
@@ -0,0 +1,86 @@
+	SCMI_ERR_GENERIC = -8,	/* Generic Error */
+	SCMI_ERR_HARDWARE = -9,	/* Hardware Error */
+	SCMI_ERR_PROTOCOL = -10,/* Protocol Error */
+	SCMI_ERR_MAX = -11
```
the MAX feels a bit weird here since it's actually the smallest number.

etienne-lms replied:
to right, i'll change `SCMI_ERR_MIN`.

apalos commented on include/scmi_agent.h:
```diff
@@ -0,0 +1,86 @@
+ */
+static inline int scmi_to_linux_errno(int errno)
+{
+	if (errno <= SCMI_SUCCESS && errno > SCMI_ERR_MAX)
```
So here the errno > SCMI_ERR_MAX is a  bit difficult to read if you are not aware of the rest of the code. 
I'd just change the SCMI_ERR_MAX to something that would make more sense

etienne-lms replied:
yes, `SCMI_ERR_MIN` seems more consistent.

jenswi-linaro replied:
I'd rather skip `SCMI_ERR_MAX` completely and check against `ARRAY_SIZE(scmi_linux_errmap)`

apalos commented on drivers/misc/scmi_agent.c:
```diff
@@ -0,0 +1,437 @@
+ * param[0] (in value) - value.a: SCMI channel identifier
+ */
+#define TA_CMD_GET_CHANNEL_ID		0x1
+#define TA_CMD_PROCESS_CHANNEL          0x2
```
Use BIT(x) on both of those, since you are already using it earlier in the code?

etienne-lms replied:
These are IDs, not  bit masks. These are command IDs send between client (here U-Boot) and the SCMI PTA in OP-TEE.

apalos replied:
ok that's fine then

etienne-lms commented on include/scmi_agent.h:
```diff
@@ -0,0 +1,86 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (C) 2019, Linaro Limited
```
Part of this code is copied from the SCMI Linux driver, released by Arm. I guess I should add `Copyright (C) 2018 ARM Ltd.`.

etienne-lms committed:
firmware: add new driver for SCMI firmwares

This change introduces SCMI agent driver in U-Boot in the firmware
U-class.

SCMI agent driver is designed for platforms that embed a SCMI server in
a firmware hosted for example by a companion co-processor or the secure
world of the executing processor.

SCMI protocols allow an SCMI agent to discover and access external
resources as clock, reset controllers and many more. SCMI agent and
server communicate following the SCMI specification [1]. SCMI agent
complies with the DT bindings defined in the Linux kernel source tree
regarding SCMI agent description since v5.8-rc1.

These bindings describe 2 supported message transport layer: using
mailbox uclass devices or using Arm SMC invocation instruction. Both
use a piece or shared memory for message data exchange.

In the current state, the SCMI agent driver does not bind to any SCMI
protocol to a U-Boot device driver. Former changes will implement
dedicated driver (i.e. an SCMI clock driver or an SCMI reset controller
driver) and add bind supported SCMI protocols in scmi_agent_bind().

Links: [1] https://developer.arm.com/architectures/system-architectures/software-standards/scmi
Signed-off-by: Etienne Carriere <etienne.carriere@linaro.org>

etienne-lms committed:
reset: add reset controller driver for SCMI agents

This change introduces a reset controller driver for SCMI agent devices.
When SCMI agent and SCMI reset domain drivers are enabled, SCMI agent
binds a reset controller device for each SCMI reset domain protocol
devices enabled in the FDT.

SCMI reset driver is embedded upon CONFIG_RESET_SCMI=y. If enabled,
CONFIG_SCMI_AGENT is also enabled.

SCMI Reset Domain protocol is defined in the SCMI specification [1].

Links: [1] https://developer.arm.com/architectures/system-architectures/software-standards/scmi
Signed-off-by: Etienne Carriere <etienne.carriere@linaro.org>

etienne-lms committed:
dt-bindings: arm: SCMI bindings documentation

Dump SCMI DT bindings documentation from Linux kernel source
tree v5.8-rc1.

Signed-off-by: Etienne Carriere <etienne.carriere@linaro.org>

etienne-lms committed:
clk: add clock driver for SCMI agents

This change introduces a clock driver for SCMI agent devices. When
SCMI agent and SCMI clock drivers are enabled, SCMI agent binds a
clock device for each SCMI clock protocol devices enabled in the FDT.

SCMI clock driver is embedded upon CONFIG_CLK_SCMI=y. If enabled,
CONFIG_SCMI_AGENT is also enabled.

SCMI Clock protocol is defined in the SCMI specification [1].

Links: [1] https://developer.arm.com/architectures/system-architectures/software-standards/scmi
Signed-off-by: Etienne Carriere <etienne.carriere@linaro.org>

etienne-lms commented:
Superseded by https://github.com/etienne-lms/u-boot/pull/4.
Thanks for the review comments. All addressed.
https://github.com/etienne-lms/u-boot/pull/4 is cleaner in that SCMI agent is defined in the firmware uclass, not the misc.
